ext/uri/php_lexbor.c

```diff
@@ -0,0 +1,604 @@
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result lexbor_write_fragment(uri_internal_t *internal_uri, zval *value, zval *errors)
+{
+	lxb_url_t *lexbor_uri = (lxb_url_t *) internal_uri->uri;
+	lexbor_str_t str = {0};
+
+	ZVAL_TO_LEXBOR_STR(value, str);
+
+	lxb_status_t status = lxb_url_api_hash_set(lexbor_uri, lexbor_parser, str.data, str.length);
+
+	CHECK_WRITE_RESULT(status, lexbor_uri, str, errors);
+}
+
+static zend_result lexbor_init_parser(void)
+{
+	lexbor_mraw_t *mraw = lexbor_mraw_create();
+	lxb_status_t status = lexbor_mraw_init(mraw, 4096 * 2);
+	if (status != LXB_STATUS_OK) {
+		lexbor_mraw_destroy(mraw, true);
+		return FAILURE;
+	}
+
+	lxb_url_parser_t *parser = lxb_url_parser_create();
+	status = lxb_url_parser_init(parser, mraw);
+	if (status != LXB_STATUS_OK) {
+		lxb_url_parser_destroy(parser, true);
+		return FAILURE;
+	}
+
+	lexbor_parser = parser;
+	lexbor_urls = 0;
+
+	zend_hash_init(&lexbor_property_handlers, 8, NULL, NULL, true);
+
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_SCHEME), lexbor_read_scheme, lexbor_write_scheme);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_USERNAME), lexbor_read_username, lexbor_write_username);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_PASSWORD), lexbor_read_password, lexbor_write_password);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_HOST), lexbor_read_host, lexbor_write_host);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_PORT), lexbor_read_port, lexbor_write_port);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_PATH), lexbor_read_path, lexbor_write_path);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_QUERY), lexbor_read_query, lexbor_write_query);
+	URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_FRAGMENT), lexbor_read_fragment, lexbor_write_fragment);
+
+	return SUCCESS;
+}
+
+static void *lexbor_parse_uri(const zend_string *uri_str, const void *base_url, zval *errors)
+{
+	lexbor_cleanup_parser();
+
+	lxb_url_t *url, *lexbor_base_url = NULL;
+
+	if (base_url) {
+		lexbor_base_url = (lxb_url_t *) base_url;
+	}
+
+	url = lxb_url_parse(lexbor_parser, lexbor_base_url, (unsigned char *) ZSTR_VAL(uri_str), ZSTR_LEN(uri_str));
+	fill_errors(errors);
+
+	return url;
+}
+
+static void lexbor_create_invalid_uri_exception(zval *exception_zv, zval *errors)
+{
+	object_init_ex(exception_zv, whatwg_invalid_url_exception_ce);
+
+	zend_update_property_string(
+		whatwg_invalid_url_exception_ce,
+		Z_OBJ_P(exception_zv),
+		ZSTR_VAL(ZSTR_KNOWN(ZEND_STR_MESSAGE)),
+		ZSTR_LEN(ZSTR_KNOWN(ZEND_STR_MESSAGE)),
+		"URL parsing failed"
+	);
```

Comment:
I think I would find this a little cleaner. Using a known string just to decompose it into val + length doesn't bring any benefit:

```suggestion
	zval value;
	ZVAL_STRING(&value, "URL parsing failed");
	zend_update_property_ex(whatwg_invalid_url_exception_ce, Z_OBJ_P(exception_zv), ZSTR_KNOWN(ZEND_STR_MESSAGE), &value);
	zval_ptr_dtor(&value);
```